clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/HonoKafkaConsumer.java

```diff
@@ -0,0 +1,458 @@
+                }
+            });
+        } else {
+            kafkaConsumer.partitionsAssignedHandler(this::onPartitionsAssigned);
+            kafkaConsumer.partitionsRevokedHandler(this::onPartitionsRevoked);
+        }
+    }
+
+    private Future<Void> subscribeAndWaitForRebalance() {
+        if (stopped.get()) {
+            return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "already stopped"));
+        }
+        final Promise<Void> subscribeDonePromise = subscribeDonePromiseRef
+                .updateAndGet(promise -> promise == null ? Promise.promise() : promise);
+        final Promise<Void> subscriptionPromise = Promise.promise();
+        if (topicPattern != null) {
+            kafkaConsumer.subscribe(topicPattern, subscriptionPromise);
+        } else {
+            kafkaConsumer.subscribe(topics, subscriptionPromise);
+        }
+        vertx.setTimer(WAIT_FOR_REBALANCE_TIMEOUT, ar -> {
+            if (!subscribeDonePromise.future().isComplete()) {
+                final String errorMsg = "timed out waiting for rebalance and update of subscribed topics";
+                log.warn(errorMsg);
+                subscribeDonePromise.tryFail(new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, errorMsg));
+            }
+        });
+        return CompositeFuture.all(subscriptionPromise.future(), subscribeDonePromise.future()).mapEmpty();
+    }
+
+    private void onPartitionsAssigned(final Set<TopicPartition> partitionsSet) {
+        if (log.isDebugEnabled()) {
+            log.debug("partitions assigned: [{}]", HonoKafkaConsumerHelper.getPartitionsDebugString(partitionsSet));
+        }
+        // if a topic pattern is used, kafkaConsumer.subscription() will be used to fetch the actual subscribed topics;
+        // if a fixed topic list is used instead, calling kafkaConsumer.subscription() would be of no use since it
+        // always returns the topics used in the kafkaConsumer.subscribe(Collection) call, regardless of whether they exist
+        // (kafkaConsumer.subscribe(Collection) would have auto-created the topics anyway, if "auto.create.topics.enable" is true)
+        if (topicPattern != null) {
+            updateSubscribedTopicPatternTopics();
+        } else {
+            Optional.ofNullable(subscribeDonePromiseRef.getAndSet(null))
+                    .ifPresent(Promise::tryComplete);
+        }
+        if (onPartitionsAssignedHandler != null) {
+            onPartitionsAssignedHandler.handle(partitionsSet);
+        }
+    }
+
+    private void updateSubscribedTopicPatternTopics() {
+        // update subscribedTopicPatternTopics (subscription() will return the actual topics, i.e. not the topic pattern if one is used
+        kafkaConsumer.subscription(ar -> {
+            if (ar.succeeded()) {
+                subscribedTopicPatternTopics = new HashSet<>(ar.result());
+            } else {
+                log.warn("failed to get subscription", ar.cause());
```

Comment:
I see no direct need for this here, invoking classes can log failed future exceptions to trace spans.
I see this class as rather low-level, not needing extra span parameters, like `HonoConnectionImpl` where there is also no tracing.

Reply:
In case of failed future exceptions I agree that the invoking classes can take care of. But there are also some log statements that could be quite useful to debug in case of any problem. We are already using  low-level Kafka invocations and IMHO it would be good to trace them.

Reply:
But we can't add a span parameter to each and every method. And in this specific case, the method mostly is called when there is no surrounding span context available. It doesn't make sense to create a span for each rebalance alone.